### ISSUE TITLE
GDScript 2.0: typed variables no longer nullable?

### ISSUE THREAD

**Issue by cosmicchipsocket** [godotengine/godot#67105]
### Godot version

4.0 beta 2

### System information

Windows 10, x64, standard build

### Issue description

In Godot 3, it was possible to have a typed variable be null. You could for example define a variable as null and later assign a `String`, or you could pass null to a function that takes a typed parameter such as `Material`. In Godot 4, this causes a type error.

### Steps to reproduce

The following script worked in Godot 3 (replace `instantiate` with `instance`):

```
extends Node

var next_screen_path = null

func _physics_process(_delta):
	if next_screen_path != null:
		var new_screen_res = load(next_screen_path) as PackedScene
		
		if !new_screen_res:
			push_error("No scene at " + next_screen_path)
			next_screen_path = null
			return
		
		var child = get_child(0)
		
		if child:
			child.free()
		
		var new_screen = new_screen_res.instantiate()
		add_child(new_screen)
		
		next_screen_path = null
		return

func set_screen(screen_path : String):
	next_screen_path = screen_path
```

In Godot 4, this gives `Line 7: Invalid argument for "load()" function: argument 1 should be String but is null.`

Changing line 3 to `var next_screen_path : String = null` also does not work: `Trying to assign value of type 'Nil' to a variable of type 'String'.`

Can types no longer be nullable in GDScript 2? What would it take to make them nullable again? Perhaps we need nullable syntax like in C# (`int?`), or is this simply a bug in the interpreter?

**Comment by KoBeWi** [comment]
`next_screen_path` is not typed, so it should be "nullable".

**Comment by cosmicchipsocket** [comment]
In this case, `next_screen_path` gets assigned to a `null` type, and the syntax checker complains because it wants a `String` for `load`.

As another example, my shader preloader often has function signatures that look like this:

```
func _try_add_mesh_material(material : Material)
```

I often feed in properties directly such as from `GeometryInstance.material_override`, which is a `Material` that is sometimes null. I then do the null check from inside the function. But how is it that `material_override`, a `Material`, can be null in the engine, but in my code I can't pass that into a function?

**Comment by amoriqbal** [comment]
Can we add a flag to Variant that indicates if the variant object is explicitly set to null?

**Comment by KoBeWi** [comment]
Object-based variants (Object, RefCounted, Resource, Node, etc.) can be null. `var something: Node = null` is a valid code. You can even have a method argument like that and pass nulls to methods that take Object-based arguments.

Non-object variants (String, int, Array, etc.) can't be null. You can assign a null to a variable that is not typed though.
`var something = null` - valid
`var something: String = null` - not valid

There is a proposal to make everything nullable: https://github.com/godotengine/godot-proposals/issues/162

**Comment by cosmicchipsocket** [comment]
This issue is not intended to be about nullable basic types. If String is in fact a basic type, I can work around that by using an empty string instead. However, `Material` is a `Resource` so I would expect that the type system would allow me to pass or assign null where the receiving type is `Material`. The fact that I cannot do this appears to be a bug.

**Comment by KoBeWi** [comment]
`Material` should already be nullable. Do you have some example where it doesn't work? (your last example doesn't seem to have enough details)

**Comment by cosmicchipsocket** [comment]
Project for Godot 3.5.0:
[gdscript_null_issue_godot3.zip](https://github.com/godotengine/godot/files/9742975/gdscript_null_issue_godot3.zip)

Project ported to Godot 4.0 beta 2, with small manual script change to make the ray pick work:
[gdscript_null_issue_godot4.zip](https://github.com/godotengine/godot/files/9742976/gdscript_null_issue_godot4.zip)

Notice how the left sphere can be clicked on in Godot 3, but in 4, the function call errors out with:
```
Invalid type in function 'update_material_text' in base 'Node ()'. The Object-derived class of argument 1 (null instance) is not a subclass of the expected argument class.
```

**Comment by KoBeWi** [comment]
The problem is with property hint. This works: `update_material_text(null)`. But your code fails, because mesh's material is defined like this:
https://github.com/godotengine/godot/blob/104af08e9d7c8308c2dd6cad89a0caee774043a0/scene/resources/primitive_meshes.cpp#L222
It has double type hint, so the property's is not recognized properly. Previously we had a reverse issue - you couldn't assign material, because it expected literally `BaseMaterial3D,ShaderMaterial`, which is not a real type.

**Comment by cosmicchipsocket** [comment]
Decided to test this theory. Looks like the issue still occurs with a single type hint. Here it is failing on `TextMesh.font`:

[gdscript_null_issue_2.zip](https://github.com/godotengine/godot/files/9749667/gdscript_null_issue_2.zip) (Godot 4.0 beta 2)

**Comment by cosmicchipsocket** [comment]
Hmm, is there something about how GDScript handles null that I'm unaware of? Both of these variables display as `<null>` in the debugger:

```
var font : Font = object.mesh.font
var null_font : Font = null
```

Are null values typed to a given class maybe?

**Comment by KoBeWi** [comment]
Yes, a `null` can either be `Variant::NIL` or `Variant::OBJECT` with `nullptr` as object pointer.

**Comment by cosmicchipsocket** [comment]
Ah, so I'm guessing either the engine type Object isn't using `Variant::OBJECT` null internally, or `Variant::NIL` isn't coercing to `Variant::OBJECT`.

Would this mean that normally, `nullptr`s for two different Object classes would be considered type compatible?

**Comment by cosmicchipsocket** [comment]
Looks like this is fixed by #62667.

**Comment by akien-mga** [comment]
Fixed by #73544.
